### ISSUE TITLE
Host 'xxx.xxx.xxx.xxx' is not allowed to connect to this MySQL server

### ISSUE THREAD

**Issue by hardikj** [docker/compose#1614]
I get the following error when my rails app try to connect to mysql
'Host 'xxx.xxx.xxx.xxx' is not allowed to connect to this MySQL server'
What am I doing wrong?

docker-compose.yml
```
db:  
  image: "mysql:5.5"
  volumes:
    - /var/lib/mysql/:/var/lib/mysql/
  expose:
    - 3306
web:
  image: "railsapp:dockerfile"
  command: /bin/bash -l -c "bundle exec rails s -p 3000 -b '0.0.0.0'"
  volumes:
    - .:/home/railsApp
  ports:
    - "3000:3000"
  links:
    - "db:db"
```

**Comment by hardikj** [comment]
This doesn't seems to be a compose issue so closing this

**Comment by mcgeeco** [comment]
Did you find out what was the cause of this?

**Comment by hardikj** [comment]
@mcgeeco, 

In my case web container was trying to connect to db container through root user. But In mysql do not allow you to connect though root user remotely. So I work around is to change the privileges  of your root user in mysql container. or you can try connecting through different user.

If you are using the same mysql image as I am you can specify  MYSQL_USER, MYSQL_PASSWORD, MYSQL_DATABASE as environment variable in your docker compose file. So mysql image will create a user and a database accessible to that user. Also, keep make sure the host that your 'web' container is trying to connect to is 'db' and not 'localhost' (at least in my case, since I have not exposed ports to my local computer)

I hope this helps

**Comment by mcgeeco** [comment]
Thanks for the reply! I've been using the env variables like that but it's still giving me that error. I have a feeling something else is going on unfortunately. I'll keep looking...

**Comment by aisensiy** [comment]
@mcgeeco I got the same problem, so why was that happened? Have you fixed that?

**Comment by mcgeeco** [comment]
Like I said the problem was that the server container was trying to access the database before the mysql server container had enough time to start up properly. The only way I could fix it was by starting the web container using a shell script that pinged the mysql server until it got a response.  It was pretty hacky...

```
#!/bin/bash

until mysqladmin ping -h db; do
    echo "$(date) - waiting for mysql"
    sleep 3
done

# Do server stuff here now that we know the mysql server is up and running on that container.
```

(in my case my mysql container was named 'db')

**Comment by aanand** [comment]
@mcgeeco That's not hacky, in my opinion - it's best practice to make your application containers resilient to database unavailability.

**Comment by laggingreflex** [comment]
I too was facing this. Initially I tried @mcgeeco's method and it worked! But then the errors became something about low disk space so I deleted all containers/images/volumes even the machine itself, and basically redid everything from scratch again, and then it worked even without @mcgeeco's method! 

I came across these similar issues: ([1](https://github.com/docker-library/mysql/issues/110), [2](https://github.com/docker-library/mysql/issues/152)), and much like them, it solved itself seemingly on its own.
